assets/js/base/context/cart-checkout/payment-methods/reducer.js

```diff
@@ -107,10 +107,7 @@ const reducer = (
 		case SET_REGISTERED_PAYMENT_METHODS:
 			return {
 				...state,
-				paymentMethods: {
-					...state.paymentMethods,
-					...paymentMethods,
-				},
+				paymentMethods,
```

Comment:
Now we're replacing/resetting the `paymentMethods` array whenever the available options (gateways) change.

Previously this action was appending - so this is a change in behaviour. I'm hoping this is fine, since the action name is `SET_` not `ADD_` - this implementation seems consistent. 

Without this change, COD will appear when you select a COD-compatible shipping option, but **won't disappear** when you select a shipping option that doesn't support it :)

@nerrad  - We could change this and track the "availability" of each method using a boolean flag in state. That's a bigger change but might be more appropriate.

Reply:
Moved across to  #2840  :)